### DISCUSSION TITLE
Auto-conversion to int for controller arguments

### DISCUSSION THREAD

**Discussion by umulmrum** [symfony/symfony#50073] [Q&A] [answered]
Hi,

in a controller, I type-hinted an argument as int.

`public function foo(Request $request, int $someId): Response`

Also, the controller uses strict typing by calling `declare(strict_types=1)`.

When the controller is called, an argument is always passed as a string containing the number, e.g. `"42"`, making the PHP type check fail.

I found that the value is resolved by `RequestAttributeValueResolver`, but despite having the `int` type available in the argument metadata, this class always returns the default string-typed value by calling `$request->attributes->get`.

So my question is: Is primitive-type-conversion really unsupported yet, or is there anything wrong with my approach? I saw that there is a recent PR for 6.3 that deals with argument conversion: https://github.com/symfony/symfony/pull/49134 - still, the programmer's intention looks clear enough to me that is seems overkill to be forced to use an attribute.

Thanks for pointers as where to go from here, thanks! :-)

**Comment by umulmrum** [comment, marked as answer]
Ah just found out what the problem is - myself of course ;-)

I don't call the controller directly, but with a detour over another class. In that class, strict types are also used. And as the caller determines if types are really strict, it then applies and makes the call fail. When using the `HttpKernel` normally, type coercion works as "expected" as it doesn't declare strict types.

Still this could solved more cleanly, but my problem vanished, so I'm happy with that :-)

**Reply by stof** [reply]
Well, Symfony itself does not declare strict types on purpose. The default coercion rules (especially in recent PHP versions) are actually _more_ type safe than using `strict_types` and then doing explicit `(int)` casts in the code (as those will happily loose data while casting: https://3v4l.org/NERuj shows this for an `int` type).

**Comment by stof** [comment]
Note that https://github.com/symfony/symfony/pull/49134 is not about dealing with parameter conversion. It is about mapping the query string to arguments.
It will rely on the various typed getters of the InputBag to avoid ending up with TypeError based on user-provided values (the query string is entirely in the control of the client, contrary to request attributes which are in the control of the routing layer)
